usePackageJsonVersion should generate a valid semver version

Currently the `usePackageJsonVersion` simply appends the build counter to the version from package.json. This means you end up with a build number in the format _Major.Minor.Patch.Counter_ (plus whatever build metadata based on the branch/commit). See https://github.com/nice-digital/teamcity-build-number/blob/master/src/index.js#L39-L40.

This isn't a valid npm SemVer version number (although it is valid with NuGet). This means you can't update the package json with this version, or use it for npm commands like `npm pack` etc. I think though it's valid expectation that using package.json to generate the build number should also result in an npm-semver-compatible version.

You get errors like:

```
npm ERR! Invalid version: "1.0.0.286-rBAF45D4"
```

I propose do one of 3 things to make the generated version valid semver:

1. use the build counter as the patch when `usePackageJsonVersion`  is `true`, overriding the patch value from the package.json
2. use a dash (instead of a dot) to separate the build counter so it becomes part of the meta data e.g. _1.2.3+286.rBAF45D4_  or _1.0.0-298.GN-52-CKSGatsbyv2_
3. append the build counter after the other build metadata e.g. _1.2.3+rBAF45D4.286_ or _1.0.0-GN-52-CKSGatsbyv2.298_

Option 1 disregards whatever is in the patch value for package json, but means we don't have to manually remember to increment patch each time. If you were to add a feature and move from 1.0.whatever to 1.1.whatever you'd potentially jump from 1.0.234 to 1.1.235 which is a bit odd as it implies there are 235 patch releases in version 1.1 which isn't the case!

Option 2, using a dash, actually means that it's treated as a pre-release, so not sure this is the right approach.

Options 3 seems good, however, npm would treat _1.2.3+rBAF45D4.286_ and _1.2.3+rABC123.287_ as the _same_ version as it disregards build metadata (after the plus) so would require manual update to increment major/minor/patch, which, let's be honest we're going to forget to do.

So, on balance, option 1 is probably best. @imranazad @John-Williams  @wa-rren-dev I'd appreciate your thoughts please if you've got a moment! Maybe there's another options I haven't considered?

Run the following to check a version on the command line:

```
npx semver 1.2.3+rBAF45D4.286 || echo "Not valid semver"
npx semver 1.0.0-GN-52-CKSGatsbyv2.298 || echo "Not valid semver"
```

--- a/src/index.js
+++ b/src/index.js
@@ -216,10 +216,18 @@ function nameMatchesConvention(enforceNamingConvention, namingConventionRegEx, n
 
 function setPackageJsonVersion(version, buildCounter)
 {
-	version = version.split(".");
-	delete version[2];
+	if(version.includes("alpha"))
+	{
+		const regex = /\d+(?=-\w+)/;
+		version = version.replace(regex, buildCounter);
 
-	return `${ version.join(".") }${ buildCounter }`;
+	} else {
+		version = version.split(".");
+		delete version[2];
+		version = `${ version.join(".") }${ buildCounter }`;
+	}
+
+	return version;
 }
 
 module.exports = {

--- a/src/index.js
+++ b/src/index.js
@@ -37,7 +37,7 @@ function setBuildNumber(usePackageJsonVersion, branch, gitHubToken, gitHubRepo, 
 		console.log(`Build counter is '${ buildCounter }'`);
 
 		// Use Major.Minor.Patch.Counter format
-		buildNumber = `${ version }.${ buildCounter }`;
+		buildNumber = setPackageJsonVersion(version, buildCounter);
 
 		console.log("##teamcity[blockClosed name='package.json version']");
 	}
@@ -214,6 +214,14 @@ function nameMatchesConvention(enforceNamingConvention, namingConventionRegEx, n
 	return namingConventionRegEx.test(nameToTest);
 }
 
+function setPackageJsonVersion(version, buildCounter)
+{
+	version = version.split(".");
+	delete version[2];
+
+	return `${ version.join(".") }${ buildCounter }`;
+}
+
 module.exports = {
 	setBuildNumber: setBuildNumber,
 	getPullRequest: getPullRequest,
@@ -223,6 +231,7 @@ module.exports = {
 	MaxBranchNameLength: MaxBranchNameLength,
 	getPackagePath: getPackagePath,
 	nameMatchesConvention: nameMatchesConvention,
+	setPackageJsonVersion: setPackageJsonVersion,
 	BranchNamingConventionRegex: BranchNamingConventionRegex,
 	PullRequestTitleNamingConventionRegex: PullRequestTitleNamingConventionRegex,
 	PullRequestRegex:PullRequestRegex

--- a/test/test.js
+++ b/test/test.js
@@ -25,6 +25,16 @@ describe("TeamCity build number", function() {
 			assert.equal("33", pull.match(index.PullRequestRegex)[2]);
 		});
 	});
+
+
+	describe("usePackageJsonVersion", function() {
+		it("should set the correct version for patch", function() {
+			var version = "0.1.0";
+			var buildCounter = "1234";
+
+			assert.equal("0.1.1234", index.setPackageJsonVersion(version, buildCounter));
+		});
+	});
 /*
 	describe("Get package.json path", function() {
 		var processCwd = "C:\\_src\\teamcity-build-number";

--- a/test/test.js
+++ b/test/test.js
@@ -16,23 +16,30 @@ describe("TeamCity build number", function() {
 	describe("pull request regex", function() {
 		it("should match merge pull request naming convention", function() {
 			var merge = "33/merge";
-			
+
 			assert.equal("33", merge.match(index.PullRequestRegex)[1]);
 		});
 		it("should match pull request naming convention", function() {
 			var pull = "pull/33";
-			
+
 			assert.equal("33", pull.match(index.PullRequestRegex)[2]);
 		});
 	});
 
 
 	describe("usePackageJsonVersion", function() {
-		it("should set the correct version for patch", function() {
+		it("should set the correct version for patch using release", function() {
 			var version = "0.1.0";
 			var buildCounter = "1234";
 
 			assert.equal("0.1.1234", index.setPackageJsonVersion(version, buildCounter));
+		});
+
+		it("should set the correct version for patch using pre-release", function() {
+			var version = "1.2.1234-alpha.0";
+			var buildCounter = "1235";
+
+			assert.equal("1.2.1235-alpha.0", index.setPackageJsonVersion(version, buildCounter));
 		});
 	});
 /*
@@ -56,7 +63,7 @@ describe("TeamCity build number", function() {
 	});
 
 	describe("Name Matches Convention", function() {
-		
+
 		it("when enforceNamingConvention is not set, true is always returned", function() {
 			var returnValue = index.nameMatchesConvention(undefined, null, null);
 			assert.equal(returnValue, true);

